ui/article-aem-renderer/src/main/java/org/cru/godtools/article/aem/ui/ArticleWebViewClient.kt

```diff
@@ -45,39 +46,31 @@ internal class ArticleWebViewClient @Inject constructor(
     }
 
     @WorkerThread
-    private fun Uri.getResponseFromFile(context: Context): WebResourceResponse? {
-        val resource = getResource() ?: return null
+    private fun Uri.getResponseFromFile(): WebResourceResponse? {
+        val resource = findResource(this) ?: return null
         val type = resource.contentType
-        val data = resource.getData(context) ?: return null
+        val data = resource.getData() ?: return null
         return WebResourceResponse(
             type?.let { "${type.type()}/${type.subtype()}" } ?: "application/octet-stream",
             type?.charset()?.name(), data
         )
     }
 
-    private fun Uri.getResource(): Resource? {
-        val resource = resourceDao.find(this) ?: return null
-        if (resource.isDownloaded) return resource
+    private fun findResource(uri: Uri) = runBlocking {
+        val resource = resourceDao.find(uri) ?: return@runBlocking null
+        if (resource.isDownloaded) return@runBlocking resource
 
         // attempt to download the file if we haven't downloaded it already
-        try {
-            // TODO: this may create a memory leak due to the call stack holding a reference to a WebView
-            aemArticleManager.enqueueDownloadResource(resource.uri, false).get()
-        } catch (e: InterruptedException) {
-            // propagate thread interruption
-            Thread.currentThread().interrupt()
-            return null
-        } catch (e: ExecutionException) {
-            Timber.tag(TAG).d(e.cause, "Error downloading resource when trying to render an article")
-        }
+        // TODO: this may create a memory leak due to the call stack holding a reference to a WebView
+        aemArticleManager.downloadResource(resource.uri, false)
 
         // refresh resource since we may have just downloaded it
-        return resourceDao.find(this)
+        return@runBlocking resourceDao.find(uri)
     }
 
-    private fun Resource.getData(context: Context) =
+    private fun Resource.getData() = runBlocking {
         try {
-            getInputStream(context)
+            getInputStream(aemFileManager)
```

Comment:
Could you explain why we no longer have to pass the context to this method and the methods above? I believe it has to do with the aemFileManager but just want to make sure.

Reply:
encapsulation, before we were resolving the location of a resource directly from the context object using static methods. This makes it pretty hard to test anything that interacted with files because you always had to mock quite a bit about the Context interactions.

By encapsulating those interactions within a `FileManager` object, we can provide a mock `FileManager` that will return a file/input stream as necessary without having to setup the actual file system interactions.

It also allows us to only perform some of the intermediate logic a single time.

The actual `FileSystem` object internally still has the `context` object to locate the file on disk